src/QirRuntime/lib/QIR/arrays.cpp

```diff
@@ -9,13 +9,29 @@
 #include <string.h> // for memcpy
 #include <vector>
 
+#include "allocationsTracker.hpp"
+#include "CoreTypes.hpp"
+#include "context.hpp"
 #include "qirTypes.hpp"
 #include "quantum__rt.hpp"
 
-#include "CoreTypes.hpp"
+namespace Microsoft
+{
+namespace Quantum
+{
+    extern thread_local std::unique_ptr<QirExecutionContext> g_context;
+}
+} // namespace Microsoft
+
+using namespace Microsoft::Quantum;
 
 int QirArray::AddRef()
 {
+    if (g_context != nullptr && g_context->trackAllocatedObjects)
+    {
+        g_context->allocationsTracker->OnAddRef(this);
+    }
+
     assert(this->refCount != 0 && "Cannot resurrect released array!");

```

Comment:
Minor:
It probably makes sense to execute the (green) line 35 before lines 30 - 34.

Reply:
I want to be able to test the tracking error handling and if I move the assert up, it would fail the tests in debug builds. The thinking here is that the tracking, when enabled, is guaranteed to detect (and handle in some well-defined way) the reference counting bugs, and the assert is a "best effort" debug-only diagnostics (and documentation).